content/6_Advanced/Offline_Del.mdx

```diff
@@ -1,100 +1,285 @@
 
-**DSU with rollback** is a subproblem required to solve the above task.
+We can now construct a query tree. If our interval is encapsulated by the the tree's interval, then we can add our query to the node corresponding to the interval. When answering queries, as we enter the interval, we can process all the operations inside the interval. As we exit the interval, we need to undo them. If we are at a leaf, we can answer type $3$ queries since we have processed all queries outside this interval $[i, i]$. Since we are processing intevals by halves each time, the depth is at most $\log n$, similar to divide and conquer.
 
-<Problems problems="rollback" />
+See the code below for implementation details. Note that similar to unite operations, we can also perform and undo operations of type $2$!
 
-<IncompleteSection>
+<LanguageSection>
+<CPPSection>
 
-explanation? check Guide to CP?
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
 
-</IncompleteSection>
+// BeginCodeSnip{DSU}
+class DSU {
+  private:
+	vector<ll> p, sz, sum;
+	// stores all history info related to merges
+	vector<pair<ll &, ll>> history;
 
-<Warning title="Watch Out!">
+  public:
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n, 1);
+		sum.resize(n);
+		iota(p.begin(), p.end(), 0);
+	}
 
-Because path compression is amortized, it does not guarauntee
-$\mathcal{O}(N \log^2 N)$ runtime. You _must_ use merging by rank.
+	void init_sum(vector<ll> &a) {
+		for (int i = 0; i < (int)a.size(); i++) { sum[i] = a[i]; }
+	}
 
-</Warning>
+	int get(int x) { return (p[x] == x) ? x : get(p[x]); }
 
-### Implementation
+	ll get_sum(int x) { return sum[get(x)]; }
 
-<LanguageSection>
+	void unite(int a, int b) {
+		a = get(a);
+		b = get(b);
+		if (a == b) return;
+		if (sz[a] < sz[b]) { swap(a, b); }
 
-<CPPSection>
+		// add to history
+		history.push_back({p[b], p[b]});
+		history.push_back({sz[a], sz[a]});
+		history.push_back({sum[a], sum[a]});
 
-<!-- verified at https://codeforces.com/contest/1217/submission/62335472 -->
+		p[b] = a;
+		sz[a] += sz[b];
+		sum[a] += sum[b];
+	}
 
-```cpp
-int p[MN], r[MN];
-int *t[MN * 40], v[MN * 40], X;
-int setv(int *a, int b) {
-	if (*a != b) t[X] = a, v[X] = *a, *a = b, ++X;
-	return b;
+	void add(int x, ll v) {
+		x = get(x);
+		history.push_back({sum[x], sum[x]});
+		sum[x] += v;
+	}
+
+	int snapshot() { return history.size(); }
+
+	void rollback(int until) {
+		while (snapshot() > until) {
+			history.back().first = history.back().second;
+			history.pop_back();
+		}
+	}
+};
+// EndCodeSnip
+
+const int MAXN = 3e5 + 1;
+
+DSU dsu(MAXN);
```

Comment:
ok honest question here
can you not just initialize this locally

Reply:
ok honest response here
does it matter?